libqdkim.man
.TH libqdkim 8 "" openqmail
.SH NAME
libqdkim \- handles DKIM signatures from stdin
.SH SYNOPSIS
\fBlibqdkim\fR \fIoptions [tags] <infile> [<privkeyfile> <outfile>]\fR

\fIoptions\fR is a series of getopt-style options to handle the behavior of
\fBlibqdkim\fR. \fItags\fR is a series like \fIoptions\fR to specify values of
tags of a signature.
.SH DESCRIPTION
\fBlibqdkim\fR creates DKIM signatures and add them to email headers or verifies
DKIM signatures. Line endings have to be \fBCRLF\fR for verification.
.SH OPTIONS
.TP
-s
." \fIkey\fR
signs a message using \fIprivkeyfile\fR, which is a path to a file with a
private key (PEM format). The base name of the file is used as selector. Reads
the email message on stdin and prints the \fIDKIM-Signature\fR header.
.TP
-v
Verifies a message and prints the result. On failure it exits with a non-zero exit code and includes the code to the result.
.TP
-h
prints short help and exists
.SH TAGS
.TP
-l
include the body length tag, if not provided it will not be included
.TP
-q
include query method tag, if not provided it will not be included
.TP
-t
include a timestamp tag, if not provided it will not be included
." .TP
." -f
." issue error if not all message's From headers are in signature
." .TP
." -h
." include Copied Headers
." .TP
." -p \fIssp\fR | \fIadsp\fR
." 0 - disable practice (default), 1- SSP, or 2 - ADSP verification
.TP
-b<\fIstandard\fR>
standard is a numeric value (1=allman, 2=ietf or 3=both)
.TP
-c<\fIcanonicalization\fR>
canonicalization is one of r, s, t or u (r=relaxed [DEFAULT], s=simple, t=relaxed/simple, u=simple/relaxed)
-d \fIdomain\fR
the domain tag, if not provided, determined from the sender/from header
.TP
-i<\fIidentity\fR>
the identity (usually sender address), if not provided it will not be included
.TP
-x<\fIexpire_time\fR>
the expire time in seconds since epoch ( DEFAULT = current time + 604800). If set to - then it will not be included
.TP
-z<\fIhash\fR>
a numeric value to set the rsa key type (1=sha1, 2=sha256, 3=both)
.TP
-y<\fIselector\fR>
the selector tag (DEFAULT=default)
.SH EXAMPLES
To sign a message and assuming your domainkeys are in \fIQMAILHOME/etc/dkimkeys\fR do:
.EX
    libqdkim -ydefault -d"example.com" -i"me@example.com" -l -b2 -ct -z2 \\
             -s "$InMsg" QMAILHOME/etc/dkimkeys/example.org/default "$OutMsg" 2>/dev/null
.EE

To verify a message, where message is a DKIM signed email:

.EX
    libqdkim -v <message>
.EE

.SH "SEE ALSO"
qmail-sdkim(8), qmail-vdkim(8)


CHANGELOG
commit f0246dee95471079078bc0cb9b1231ab7df20dab
Author: Kai Peter <kp@openqmail.org>
Date:   Sat May 13 00:02:48 2017 +0200

    removed dependancy 'unix2dos'

commit 258ab6af23bcb6506684566ab2d60dc7c8201dfd
Author: Kai Peter <kp@openqmail.org>
Date:   Fri May 12 09:57:32 2017 +0200

    dkimsign: corrected strings for c= tag

commit a57bafd4056759ca48a2f33aac170618a58cda03
Author: Kai Peter <kp@openqmail.org>
Date:   Thu May 11 19:20:14 2017 +0200

    cleanup dns.h

commit 946c7e207d39881d2bbc456783d9ade016e5975a
Author: Kai Peter <kp@openqmail.org>
Date:   Thu May 11 19:16:07 2017 +0200

    upped version to 20170511-beta1

commit 2997c025177f496e16448ce09bd727a1660aecd6
Merge: e7b3213 c3a5226
Author: Kai Peter <kp@openqmail.org>
Date:   Thu May 11 16:00:06 2017 +0200

    Merge pull request #2 from kp-org/dev
    
    merge of 'qdkim' into 'xdkim'

commit c3a5226ce8edcebfcc17481fb3be2a451f3fa061
Author: Kai Peter <kp@openqmail.org>
Date:   Thu May 11 15:55:16 2017 +0200

    clean-up, docs

commit ab48b27dcacb3ed370b556536915ff6b4220a07b
Author: Kai Peter <kp@openqmail.org>
Date:   Thu May 11 15:50:30 2017 +0200

    added libqdkim code

commit e47d1b44a28aa3f9bd009e23c832ab5613e25635
Author: Kai Peter <kp@openqmail.org>
Date:   Thu May 11 15:49:03 2017 +0200

    added config file

commit b49457386bfbfad3317968026e70697bff326fc6
Author: Kai Peter <kp@openqmail.org>
Date:   Wed May 10 10:19:25 2017 +0200

    qmail-sdkim: removed obsolete code, updated man page

commit 2f3a0d5a2f09982a1a606ef95bd5833298043f55
Author: Kai Peter <kp@openqmail.org>
Date:   Wed May 10 09:56:49 2017 +0200

    renamed 'mkdomainkey' to 'mkdkimkey', improved code and man page

commit 3fb114135c4dd95b235b97702f20998850052115
Author: Kai Peter <kp@openqmail.org>
Date:   Sun May 7 08:57:43 2017 +0200

    qmail-sdkim: more failsafe, usage of config file

commit 09e643ca2a0d961612e89f049a294135bcc0e81e
Author: Kai Peter <kp@openqmail.org>
Date:   Fri May 5 11:25:02 2017 +0200

    added Makefile and switched to 'make' build process

commit c579989a7e34e98c7cd5153bd7254eceaa6ba962
Author: Kai Peter <kp@openqmail.org>
Date:   Fri May 5 11:07:02 2017 +0200

    typo in mkdomainkey.8

commit e7b3213cac84b9a3be220680b2bf9878e98cd3a1
Merge: 939df0d 4cac84d
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Feb 19 11:47:42 2016 +0100

    Merge pull request #1 from kp-org/dev
    
    first merge to master

commit 4cac84d5267b25b23d6021274e3f4273fd69a2ea
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Feb 19 11:13:44 2016 +0100

    initial commit

commit 939df0d2726c2ea71c805d5cb150f1231aa3630e
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Feb 18 12:59:10 2016 +0100

    Create README.md


########## BELOW IS THE CHANGELOG BEFORE USING GIT ##########

Changes: qmail-xdkim ****************************************

20170219: created git repo
20160218: qmail-vdkim-0.42-beta
          - added variable XDKIM
          - created man page qmail-vdkim.8
20160211: qmail-sdkim-0.44
          - added variable XDKIM
20160210: qmail-sdkim-0.43
          - added PARENTLEVEL for third and fourth level domains

Changes: libdkim/xdkim **************************************

20160426:
  - changed license to ISC
20160211:
  - removed unused DKIMSignGetSig function (commented out)
  - added man page
  - deleted WIN32 stuff
20160210: converted all files to UNIX format
  - added configure, Makefile.in
  - removed '#ifdef WIN32' (commented out)
  - changed default selector to "default"
  - unset "bardenhagen" identity and domain
20160208:
  - disabled requirement 'subject' (changed options to 1)
20160205:
  - strtok_r.patch
  - fix_warnings.patch
  - dos2unix Makefile
  - added '-lresolv' to LDFLAGS
  - applied corrected usage patch


libqdkim.cpp
/*****************************************************************************
*  Copyright 2005 Alt-N Technologies, Ltd. 
*
*  Licensed under the Apache License, Version 2.0 (the "License"); 
*  you may not use this file except in compliance with the License. 
*  You may obtain a copy of the License at 
*
*      http://www.apache.org/licenses/LICENSE-2.0 
*
*  This code incorporates intellectual property owned by Yahoo! and licensed 
*  pursuant to the Yahoo! DomainKeys Patent License Agreement.
*
*  Unless required by applicable law or agreed to in writing, software 
*  distributed under the License is distributed on an "AS IS" BASIS, 
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
*  See the License for the specific language governing permissions and 
*  limitations under the License.
*
*****************************************************************************/
#define strnicmp strncasecmp 

#include <stdio.h>
#include <string.h>
#include <time.h>
#include <stdlib.h>
#include "dkim.h"
#include "dns.h"

// change these to your selector name, domain name, etc
#define MYSELECTOR	"default"
#define MYDOMAIN	"" //"bardenhagen.com"
#define MYIDENTITY	"" //"dkimtest@bardenhagen.com"

int DKIM_CALL SignThisHeader(const char* szHeader)
{
	if( strnicmp( szHeader, "X-", 2 ) == 0 )
	{
		return 0;
	}
	return 1;
}


int DKIM_CALL SelectorCallback(const char* szFQDN, char* szBuffer, int nBufLen )
{
	return 0;
}

void usage()
{
  char* version;
  version = (char *)DKIMVersion();
  printf("libqdkim %s \n",version);
  printf("Usage: libqdkim [-h|-v|-s] [tags] <msgfile> [<privkeyfile> <outfile>]\n\n");
  printf( "Options:\n\t-h show this help\n");
  printf( "\t-s sign the message\n");
  printf( "\t-v verify the message\n\n");
  printf( "These tags are available:\n");
  printf( "\t-b<standard>         - set the standard (1=allman, 2=ietf, 3=both)\n");
  printf( "\t-c<canonicalization> - r=relaxed [DEFAULT], s=simple, t=relaxed/simple, u=simple/relaxed\n");
  printf( "\t-d<domain>           - domain tag (if not provided it will be determined from the sender/from header)\n");
  printf( "\t-i<identity>         - the identity, usually the sender address (optional)\n");
  printf( "\t-l                   - include body length tag (optional)\n");
  printf( "\t-q                   - include query method tag\n");
  printf( "\t-t                   - include a timestamp tag (optional)\n");
  printf( "\t-x<expire_time>      - the expire time in seconds since epoch (optional, DEFAULT = current time + 604800)\n");
  printf( "\t-y<selector>         - set selector (DEFAULT: default)\n");
  printf( "\t-z<hash>             - set rsa-key type (1=sha1, 2=sha256, 3=both)\n");
}
int main(int argc, char* argv[])
{
  int n;
  const char* PrivKeyFile = "test.pem";
  const char* MsgFile = "test.msg";
  const char* OutFile = "signed.msg";
  int nPrivKeyLen;
  char PrivKey[2048];
  char Buffer[1024];
  int BufLen;
  char szSignature[10024];
  time_t t;
  DKIMContext ctxt;
  DKIMSignOptions opts = {0};

  opts.nHash = DKIM_HASH_SHA1_AND_256;

  time(&t);

  opts.nCanon = DKIM_SIGN_RELAXED;
  opts.nIncludeBodyLengthTag = 0;
  opts.nIncludeQueryMethod = 0;
  opts.nIncludeTimeStamp = 0;
  opts.expireTime = t + 604800;    // expires in 1 week
  strcpy( opts.szSelector, MYSELECTOR );
  strcpy( opts.szDomain, MYDOMAIN );
  strcpy( opts.szIdentity, MYIDENTITY );
  opts.pfnHeaderCallback = SignThisHeader;
  strcpy( opts.szRequiredHeaders, "NonExistant" );
  opts.nIncludeCopiedHeaders = 0;
  opts.nIncludeBodyHash = DKIM_BODYHASH_BOTH;

  int nArgParseState = 0;
  bool bSign = true;

  if(argc<2){
    usage();
    exit(1);
  }

	for( n = 1; n < argc; n++ )
	{
		if( argv[n][0] == '-' && strlen(argv[n]) > 1 )
		{
			switch( argv[n][1] )
			{
			case 'b':		// allman or ietf draft 1 or both
				opts.nIncludeBodyHash = atoi( &argv[n][2] );
				break;

			case 'c':		// canonicalization
				if( argv[n][2] == 'r' )
				{
					opts.nCanon = DKIM_SIGN_RELAXED;
				}
				else if( argv[n][2] == 's' )
				{
					opts.nCanon = DKIM_SIGN_SIMPLE;
				}
				else if( argv[n][2] == 't' )
				{
					opts.nCanon = DKIM_SIGN_RELAXED_SIMPLE;
				}
				else if( argv[n][2] == 'u' )
				{
					opts.nCanon = DKIM_SIGN_SIMPLE_RELAXED;
				}
				break;

			case 'd': 
				strncpy(opts.szDomain,(const char*)(argv[n]+2),sizeof(opts.szDomain)-1);
				break;
			case 'l':		// body length tag
				opts.nIncludeBodyLengthTag = 1;
				break;


			case 'h':
				usage();
				return 0;

			case 'i':		// identity 
				if( argv[n][2] == '-' )
				{
					opts.szIdentity[0] = '\0';
				}
				else
				{
					strncpy( opts.szIdentity, argv[n] + 2,sizeof(opts.szIdentity)-1 );
				}
				break;

			case 'q':		// query method tag
				opts.nIncludeQueryMethod = 1;
				break;

			case 's':		// sign
				bSign = true;
				break;

			case 't':		// timestamp tag
				opts.nIncludeTimeStamp = 1;
				break;

			case 'v':		// verify
				bSign = false;
				break;

			case 'x':		// expire time 
				if( argv[n][2] == '-' )
				{
					opts.expireTime = 0;
				}
				else
				{
					opts.expireTime = t + atoi( argv[n] + 2  );
				}
				break;

			case 'y':
				strncpy( opts.szSelector, argv[n]+2, sizeof(opts.szSelector)-1);
				break;

			case 'z':		// sign w/ sha1, sha256 or both 
				opts.nHash = atoi( &argv[n][2] );
				break;
			}
		}
		else
		{
			switch( nArgParseState )
			{
			case 0:
				MsgFile = argv[n];
				break;
			case 1:
				PrivKeyFile = argv[n];
				break;
			case 2:
				OutFile = argv[n];
				break;
			}
			nArgParseState++;
		}
	}


  if( bSign )
  {
    FILE* PrivKeyFP = fopen( PrivKeyFile, "r" );

    if ( PrivKeyFP == NULL )
    {
      printf( "libqdkim: can't open private key file %s\n", PrivKeyFile );
      exit(1);
    }
    nPrivKeyLen = fread( PrivKey, 1, sizeof(PrivKey), PrivKeyFP );
    if (nPrivKeyLen == sizeof(PrivKey)) { /* TC9 */
      printf( "libqdkim: private key buffer isn't big enough, use a smaller private key or recompile.\n");
      exit(1);
    }
    PrivKey[nPrivKeyLen] = '\0';
    fclose(PrivKeyFP);


    FILE* MsgFP = fopen( MsgFile, "rb" );

    if ( MsgFP == NULL ) 
    {
      printf( "libqdkim: can't open msg file %s\n", MsgFile );
      exit(1);
    }

    n = DKIMSignInit( &ctxt, &opts );

    while (1) {

			BufLen = fread( Buffer, 1, sizeof(Buffer), MsgFP );

			if( BufLen > 0 )
			{
				DKIMSignProcess( &ctxt, Buffer, BufLen );
			}
			else
			{
				break;
			}
    }

    fclose( MsgFP );

		//n = DKIMSignGetSig( &ctxt, PrivKey, szSignature, sizeof(szSignature) );

    char* pSig = NULL;

    n = DKIMSignGetSig2( &ctxt, PrivKey, &pSig );

    strcpy( szSignature, pSig );

    DKIMSignFree( &ctxt );

    FILE* in = fopen( MsgFile, "rb" );
    FILE* out = fopen( OutFile, "wb+" );

    fwrite( szSignature, 1, strlen(szSignature), out );
    fwrite( "\r\n", 1, 2, out );

    while (1) {
      BufLen = fread( Buffer, 1, sizeof(Buffer), in );
      if( BufLen > 0 ) {
        fwrite( Buffer, 1, BufLen, out );
      } else {
        break;
      }
    }

    fclose( in );
  }
  else
  {
    FILE* in = fopen( MsgFile, "rb" );

    DKIMVerifyOptions vopts = {0};
    vopts.pfnSelectorCallback = NULL; //SelectorCallback;

    n = DKIMVerifyInit( &ctxt, &vopts );

    while (1) {
			
			BufLen = fread( Buffer, 1, sizeof(Buffer), in );

			if( BufLen > 0 )
			{
				DKIMVerifyProcess( &ctxt, Buffer, BufLen );
			}
			else
			{
				break;
			}
    }

    n = DKIMVerifyResults( &ctxt );

    int nSigCount = 0;
    DKIMVerifyDetails* pDetails;
    char szPolicy[512];

    n = DKIMVerifyGetDetails(&ctxt, &nSigCount, &pDetails, szPolicy );

    for ( int i = 0; i < nSigCount; i++)
    {
			printf( "Signature #%d: ", i + 1 );

			if( pDetails[i].nResult >= 0 )
				printf( "Success\n" );
			else {
				printf( "Failure" );
			int e;
			e = pDetails[i].nResult;
//			printf("Details: %i, Policy: %s\n", pDetails->nResult, d);
//			printf(" (Code %i)\n", e);
			printf(" %i\n", e);
			}
    }
    DKIMVerifyFree( &ctxt );
  }
  return 0;
}
